compiler/src/iree/compiler/Dialect/HAL/Transforms/PruneExecutables.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+#define GEN_PASS_DEF_PRUNEEXECUTABLESPASS
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h.inc"
+
+namespace {
+
+struct SymbolReferences {
+  Operation *symbolOp;
+  unsigned count = 0;
+};
+using SymbolReferenceMap = DenseMap<Attribute, SymbolReferences>;
+
+static void markReferenced(SymbolRefAttr symbolRefAttr,
+                           SymbolReferenceMap &referenceMap) {
+  auto markReferencedNested = [&](StringAttr rootRefAttr,
+                                  ArrayRef<FlatSymbolRefAttr> nestedRefAttrs) {
+    auto nestedRefAttr = nestedRefAttrs.empty()
+                             ? SymbolRefAttr::get(rootRefAttr)
+                             : SymbolRefAttr::get(rootRefAttr, nestedRefAttrs);
+    auto it = referenceMap.find(nestedRefAttr);
+    if (it != referenceMap.end())
+      ++it->second.count;
```

Comment:
Should this pass be a fixed point iteration? Isn't that how SymbolDCE normally runs?

I was trying to step through what this code is doing and was wondering if `count` be decremented anywhere. I see it incremented here and then there is a check for it equaling zero below, but no decrements?

This looked sort of like an iterative approach, if it were to decrement references:

```c++
    // Erase any executable-related op with no references.
    // We have to start with exports > variants > executables so that we don't
    // erase a container before the nested ops.
    eraseOps(exportRefAttrs, referenceMap);
    eraseOps(variantRefAttrs, referenceMap);
    eraseOps(executableRefAttrs, referenceMap);
```

I also don't see any test cases for reference _chains_ that are orphaned, just single orphaned references. (Are such chains possible today, or impossible by construction?)

Reply:
executables cannot reference other executables (or anything else), so there's no need to run fixed-point - if there are no uses of an executable (or contents of an executable) in the program we can remove them and then there are no more uses.

I'll add a test for removing all.